### ISSUE TITLE
Regression: RPC call compatibility broken with 3rd party client

### ISSUE THREAD

**Issue by lvella** [transmission/transmission#6630]
### What is the issue?

I don't know if this is intentional, but in latest main, if I try to remove a torrent and delete the data from one third party client, the data is not removed.

I tracked down the issue to [this line](https://github.com/transmission/transmission/blob/7f029acf6e1057b5e81cc91905b3bdd23ea6170e/libtransmission/rpcimpl.cc#L244):
```c++
    auto const delete_flag = args_in.value_if<bool>(TR_KEY_delete_local_data).value_or(false);
```
where the value given to `TR_KEY_delete_local_data` is not properly parsed by `get_if<bool>()`, indirectly called by `value_if<bool>()`, and returns `nullptr`, which is translated to an empty option in `value_if<bool>()`.

There client affected is Transmission Remote GUI v5.18.0. Wireshark tells that is sends `"delete-local-data": 1` instead of `"delete-local-data": true` in the Json request.

### Which application of Transmission?

transmission-daemon

### Which version of Transmission?

7f029acf6e1057b5e81cc91905b3bdd23ea6170e

**Comment by killemov** [comment]
> Wireshark tells that is sends "delete-local-data": 1 instead of "delete-local-data": true in the Json request.

This probably is the problem. In JavaScript they are both truthy, but since the parsing is done in C++ a value of 1 does not get properly parsed to boolean `true`. So the client app should get the code updated to send a proper `true`, but a point could be made for a more lenient approach towards truthy values as well.

**Comment by tearfur** [comment]
Before 22cde5d4b9ac43ac73f1309cb1021f0dd187e9f0, we used to support some truthy values since we use `tr_variantDictFindBool()` to get `delete-local-data`.

I don't know if this was a intended change.

CC @ckerr as the commit author.

**Comment by fredo-47** [comment]
In a fork of transgui, this was resolved in latest build:
https://github.com/xavery/transgui/issues/79
https://github.com/xavery/transgui/commit/409512f85001d0865dba2a21ce1497b2c60c0d8d

We also reported it at original transgui and a PR was given there, but it was not processed yet.
https://github.com/transmission-remote-gui/transgui/pull/1467

**Comment by xavery** [comment]
Not a Transmission bug in my opinion. Those fields have been specified as `boolean` ever since so this is the clients' problem for not adhering to the spec.

**Comment by killemov** [comment]
From [RPC spec](https://github.com/transmission/transmission/blob/main/docs/rpc-spec.md):

<table><tr>
<td>torrent-get</td>
<td>⚠️ old arg wanted was implemented as an array of 0 or 1 in Transmission 3.00 and older, despite being documented as an array of booleans. Transmission 4.0.0 and 4.0.1 "fixed" this by returning an array of booleans; but in practical terms, this change caused an unannounced breaking change for any 3rd party code that expected 0 or 1. For this reason, 4.0.2 restored the 3.00 behavior and updated this spec to match the code.
</td>
</tr></table>

So that. Defined as booleans since forever but actually returned as 0/1's. And when the code got "fixed" to match the definition it borked several client applications because THEY weren't lenient with truthy/falsy values. JavaScript/Web applications, such as [https://github.com/killemov/Shift](Shift) should have no/less problems because the data is interpreted as truthy/falsy by default. (1 !== true though)

**Comment by reardonia** [comment]
> Not a Transmission bug in my opinion. Those fields have been specified as `boolean` ever since so this is the clients' problem for not adhering to the spec. See my original comment in [xavery/transgui#79 (comment)](https://github.com/xavery/transgui/issues/79#issuecomment-1939592753) .

Boy, would love to see you argue with Linus about "user space" compatibility.

Pretty sure the community disagree with you on this. Back-compat should always be a top priority, unless you are specifically curing a security hole.

**Comment by reardonia** [comment]
Let me restate:

the rpc spec for transmission was written *after* the code.  The spec was wrong.  This happens often.  After wide adoption of the wrongness, the spec should be rewritten.  If you'd like, we could edit the spec to say the parameter is either a bool or a string limited to either "true" or "false".  Possibly also a 0/1 numeric. 

The official JSON schema says that only true/false (unquoted) are allowed.  But 0/1 and quoted 0/1 are accepted (can be compared with boolean true/false) in every Javascript implementation I've tried.

If you want to strictly limit to well-formed JSON, then we should version the API as such.  But for legacy client code (and there is a lot) then we should maintain Tr3 compat wherever possible. 

ps: hah, if you knew my background you'd probably laugh as you explained what 'JS' stands for

**Comment by killemov** [comment]
> quoted 0/1 are accepted 

Sure they are accepted ... but they both resolve to true. And I'm also saying there is an upside to more lenient parsing.

**Comment by killemov** [comment]
> er, no. "0" is coerced to false, "1" is coerced to true.

Correct, when explicitly comparing "0" == false the "0" even gets doubly coerced (String -> Integer -> Boolean). That power lies in the `==` operator, not in the operands themselves. So if you rewrite to:
```javascript
j = { "foo": "bar", "b": "0" };
if( j['b'] ) {
  document.writeln("true");
}
else if( !j['b'] ) {
  document.writeln("false");
}
else {
  document.writeln("truthy");
}
```
You see "true". The question then is who loosely compares to true/false primitives in if statements? (I get comparing with `===` and `!==` might be useful.)

Also if you rewrite to:
```javascript
j = { "foo": "bar", "b": "false" };
if( j['b'] == true ) {
  document.writeln("true");
}
else if( j['b'] == false ) {
  document.writeln("false");
}
else {
  document.writeln("truthy");
}
```
You see "truthy" because because both "true" and "false" are not coerced to a primitive Boolean. If the primitive Boolean was coerced to String it would have worked.

In short:
```javascript
document.writeln( "1" == true );
document.writeln( "0" == false );
document.writeln( "true" == true );
document.writeln( "false" == false );
document.writeln( "1" ? true : false );
document.writeln( "0" ? true : false );
document.writeln( "true" ? true : false );
document.writeln( "false" ? true : false );
```
Yields: `true true false false true true true true` 

<b>HOWEVER</b>: There are JSON parsers out there that DO convert both "0" and "false" to false and both "1" and "true" to true. This should be configurable, which it's usually not, because sometimes "0", "1", "true", "TRUE", "false", "FALSE" are just stings. Maybe do a mind experiment with a Date in JSON ...

**Comment by tearfur** [comment]
@killemov Why are you like this? Why has this devolved into an argument about _**javascript trivia**_?? Please discuss about backwards compatibility!!?? reardonia was not addressing your comment in the first place, but xavery's. Nonetheless, you rebutted as if he was addressing you, and it went downhill from there.

Someone please hide all these comments as "offtopic", I'll have done it long ago if I had the rights to. And when you are done hide this comment too.

**Comment by killemov** [comment]
@tearfur The discussion was, IMHO, about definition, documentation, backwards compatibility and leniency (especially in other programming languages). Should the RPC in other programming languages interpret "0" as false or should they not? Even if in JavaScript Boolean("0") yields true? So it was both about backwards compatibility and getting it right.

**Comment by tearfur** [comment]
Well, then please, what point are you trying to make by elucidating all that Javascript trivia? Make your point and make it clear. What do you think is the correct path forward for Transmission?

**Comment by killemov** [comment]
@killemov wrote:
> So the client app should get the code updated to send a proper `true`, but a point could be made for a more lenient approach towards truthy values as well.

So code and document for proper unquoted true/false values and be lenient/backwards compatible with truthy (JavaScript) values. BUT "0" (quotes!) is truthy(!) so that case goes beyond lenient and the discussion is if that is to be treated as false as well.

**Comment by tearfur** [comment]
Before anything, I apologise for overreacting. Here's my 2 cents on this whole topic.

For `4.x.x`, Transmission should keep the existing lenient Boolean parsing.

Ultimately (say starting from `5.0.0`), Transmission should only be obligated to support unquoted true/false values, anything more lenient is courtesy, and should not be relied on by API users. For this, we need to:
1. Fix `torrent-get`'s `wanted` arg to return actual Boolean values.
2. Make sure to announce in `5.0.0`'s release notes and docs about this.
3. Announce this plan early, say in the next release.

@reardonia: 
> Back-compat should always be a top priority, unless you are specifically curing a security hole.

I disagree with this, we should be improving our API and eliminating quirks in it. If that means making breaking changes, so be it, that's what semver major versions are for. Especially in this case, where the impact of this change on 3rd party developers will be small.

I will agree on this *if* we are some critically-depended-on software, but we are not.

**Comment by lvella** [comment]
My opinion on this is: even if we keep the stricter parsing, it should at least log an error when getting the wrong type in the message, and not just blindly interpret it as the default value.

**Comment by killemov** [comment]
@tearfur We all have our knee-jerk moments ...

> ... we should be improving our API and eliminating quirks in it.
 
What about removing duplicate data/structures? So remove the file-count field, priorities array and wanted array and use the fileStats structure for setting as well? Sparse arrays don't travel too well in JSON, but we could use an optional indices array when setting wanted/priority for subsets of files. (That is to accommodate for setting a single file priority in a ridiculous torrent of 10k+ files.)

**Comment by killemov** [comment]
> ... we can't assume people are always going to set `wanted` and `priority` together.

These sub-fields for `set-torrent` are all optional of course. BUT there is an `and` operation between `wanted` and `priority`, the priority only has meaning after wanted.

**Comment by killemov** [comment]
In a very narrow view you could say that daemon and web-client are developed in tandem and the rest is gravy. I believe that the effort of version control at method-level is not feasible. On the client side I'd rather deal with RPC-version as a whole. See [versions of RPC in Shift](https://github.com/killemov/Shift/blob/6997f0ab3caf64198573fe07f9f0452e86928f47/shift.js#L4905). So I develop for the latest version and provide fall backs for older daemon versions based on RPC-version. On the server side it's a completely different story: only the latest version and no fall back for older clients.
